contracts/GAUC.sol

```diff
@@ -0,0 +1,364 @@
+            address(_token),
+            _ids,
+            _borrower,
+            _auctionEndTimestamp,
+            _maxFaceValue,
+            _minBidIncrement,
+            _maturity,
+            _price,
+            address(wallet)
+        );
+    }
+
+    function cancel(uint256 _auctionId) public override {
+        AuctionInfo storage auction = updateAuctionStatus(_auctionId);
+
+        require(
+            auction.auctionStatus == AUCTION_STATUS.OPEN ||
+                auction.auctionStatus == AUCTION_STATUS.EXPIRED,
+            "GAUC: auction not open or expired"
+        );
+        require(auction.lowestBidder == address(0), "GAUC: bidder exists");
+
+        GSDIWallet(auction.IGSDIWallet).setExecutor(auction.borrower);
+        auction.auctionStatus = AUCTION_STATUS.CANCELED;
+
+        emit AuctionCancel(msg.sender, _auctionId);
+    }
+
+    function bid(uint256 _auctionId, uint256 _amount) public override {
+        AuctionInfo storage auction = updateAuctionStatus(_auctionId);
+        uint256 purchasePrice = getPurchasePrice(auction.price);
+
+        require(
+            auction.auctionStatus == AUCTION_STATUS.OPEN,
+            "GAUC: auction not open"
+        );
+        require(
+            auction.minBidIncrement <= _amount,
+            "GAUC: must be at least minBidIncrement"
+        );
+        require(
+            auction.lowestBid.mul(99) >= _amount.mul(100),
+            "GAUC: must be 1% lower than lowest bid"
+        );
+        require(
+            msg.sender == auction.lowestBidder ||
+                balanceAvailable(msg.sender) >= purchasePrice,
+            "GAUC: insufficient balance"
+        );
+
+        // unlock prev bidder balance
+        balanceLocked[auction.lowestBidder] = balanceLocked[
+            auction.lowestBidder
+        ]
+            .sub(purchasePrice);
+
+        auction.lowestBid = _amount;
+        auction.lowestBidder = msg.sender;
+
+        // lock bidder balance
+        balanceLocked[msg.sender] = balanceLocked[msg.sender].add(
+            purchasePrice
+        );
+
+        emit AuctionBid(msg.sender, _auctionId, _amount);
+    }
+
+    function claim(uint256 _auctionId) public override {
+        AuctionInfo storage auction = updateAuctionStatus(_auctionId);
+
+        require(
+            auction.auctionStatus == AUCTION_STATUS.CLAIMABLE,
+            "GAUC: auction not claimable"
+        );
+        require(auction.lowestBidder == msg.sender, "GAUC: invalid claimer");
+
+        uint256 tokenId =
+            gsdiNFT.propose(
+                auction.maturity,
+                auction.lowestBid,
+                auction.price,
+                GSDIWallet(auction.IGSDIWallet),
+                dai,
+                auction.borrower
+            );
+
+        uint256 purchasePrice = getPurchasePrice(auction.price);
+        IERC20(dai).safeApprove(address(gsdiNFT), 0);
+        IERC20(dai).safeApprove(address(gsdiNFT), purchasePrice);
```

Comment:
Dai does not implement safeApprove on all networks - use approve instead.
https://etherscan.io/address/0x6b175474e89094c44da98b954eedeac495271d0f#code